src/main/java/com/palantir/refreshable/DefaultRefreshable.java

```diff
@@ -100,64 +122,91 @@ public T current() {
+            int subscribers = orderedSubscribers.size() + 1;
+            if (subscribers > WARN_THRESHOLD) {
+                log.warn(
+                        "Refreshable {} has an excessive number of subscribers: {} and is likely leaking memory. "
+                                + "The current warning threshold is {}.",
+                        SafeArg.of("refreshableIdentifier", System.identityHashCode(this)),
+                        SafeArg.of("numSubscribers", subscribers),
+                        SafeArg.of("warningThreshold", WARN_THRESHOLD),
+                        new SafeRuntimeException("location"));
+            } else if (log.isDebugEnabled()) {
+                log.debug(
+                        "Added a subscription to refreshable {} resulting in {} subscriptions",
+                        SafeArg.of("refreshableIdentifier", System.identityHashCode(this)),
+                        SafeArg.of("numSubscribers", subscribers));
+            }
         }
+    }
 
+    @GuardedBy("readLock")
+    private Disposable subscribeToSelf(Consumer<? super T> subscriber) {
+        preSubscribeLogging();
+        orderedSubscribers.add(subscriber);
         subscriber.accept(current);
-        return () -> remove(subscriber);
+        return new DefaultDisposable(orderedSubscribers, subscriber);
     }
 
-    private synchronized void remove(Consumer<? super T> subscriber) {
-        orderedSubscribers.remove(subscriber);
-    }
+    /**
+     * A {@link Disposable} object shouldn't prevent a {@link Refreshable} from being garbage collected if the
+     * refreshable is no longer referenced. In that case it's impossible that the consumer could be called with an
+     * update because the root refreshable has been garbage collected.
+     * <pre>{@code
+     * SettableRefreshable root = Refreshable.create(initial);
+     * Disposable subscription = root.subscribe(System.out::println);
+     * root = null;
+     * // root should be garbage collected, nothing can update the SettableRefreshable value.
+     * }</pre>
+     */
+    private static final class DefaultDisposable implements Disposable {
+        private final Set<? extends Consumer<?>> subscribers;
+        private final Consumer<?> subscriber;
 
-    @Override
-    public synchronized <R> Refreshable<R> map(Function<? super T, R> function) {
-        if (mapsSinceLastUpdate.incrementAndGet() > CLEAN_THRESHOLD) {
-            // iterating over a copy allows SelfRemovingSubscribers to remove themselves without
-            // ConcurrentModificationExceptions
-            ImmutableList.copyOf(orderedSubscribers).forEach(value -> {
-                if (value instanceof SelfRemovingMapSubscriber) {
-                    ((SelfRemovingMapSubscriber<?, ?>) value).disposeIfChildHasBeenCollected();
-                }
-            });
-            mapsSinceLastUpdate.set(0);
+        DefaultDisposable(Set<? extends Consumer<?>> subscribers, Consumer<?> subscriber) {
```

Comment:
The problem is that we'd need to convert the `set<consumer>` into `consumer<consumer>` which may either capture references, or produce another thin object wrapper around the same object -- I'd opt to avoid the indirection and use the types we already have.